### PR TITLE
(EML) Add exchange monthly limits compliance module

### PR DESCRIPTION
**Pull request by aliarbak**

### CODE REVIEW

**Comment by aliarbak** [comment]
TODO:
- remove onlyComplianceCall modifier from addExchangeID and removeExchangeID functions (need something like onlyOwner for tokeny wallets)

File: test/compliances/module-exchange-monthly-limits.test.ts [hunk 1]
```diff
@@ -0,0 +1,648 @@
...
+  return {
+    ...context,
+    suite: {
+      ...context.suite,
+      complianceModule,
+    },
+  };
+}
+
+describe.only('Compliance Module: ExchangeMonthlyLimits', () => {
```
Comment (aliarbak) [inline test/compliances/module-exchange-monthly-limits.test.ts:36]: Remove describe.only

File: test/fixtures/deploy-full-suite.fixture.ts [hunk 2]
```diff
@@ -252,6 +252,7 @@ export async function deploySuiteWithModularCompliancesFixture() {
 
   const complianceProxy = await ethers.deployContract('ModularComplianceProxy', [context.authorities.trexImplementationAuthority.address]);
   const compliance = await ethers.getContractAt('ModularCompliance', complianceProxy.address);
+  await compliance.bindToken(context.suite.token.address);
```
Comment (aliarbak) [inline test/fixtures/deploy-full-suite.fixture.ts:255]: Bind token to the compliance in the test itself (not in the generic fixture file)

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+ */
+
+pragma solidity 0.8.17;
+
+import "../IModularCompliance.sol";
+import "../../../token/IToken.sol";
+import "../../../roles/AgentRole.sol";
+import "./AbstractModule.sol";
+
+contract ExchangeMonthlyLimitsModule is AbstractModule {
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:70]:
```suggestion
import "@openzeppelin/contracts/access/Ownable.sol";

contract ExchangeMonthlyLimitsModule is AbstractModule, Ownable {
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    }
+
+    /// Getter for Tokens monthlyLimit
+    mapping(address => mapping(address => uint256)) private _exchangeMonthlyLimit;
+
+    /// Mapping for users Counters
+    mapping(address => mapping(address => mapping(address => ExchangeTransferCounter))) private _exchangeCounters;
+
+    /// Mapping for wallets tagged as exchange wallets
+    mapping(address => mapping(address => bool)) private _exchangeIDs;
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:84]:
```suggestion
    mapping(address => bool) private _exchangeIDs;
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+     */
+    event ExchangeMonthlyLimitUpdated(address indexed compliance, address _exchangeID, uint _newExchangeMonthlyLimit);
+
+    /**
+    *  this event is emitted whenever an ONCHAINID is tagged as being an exchange ID.
+    *  the event is emitted by 'addExchangeID'.
+    *  `compliance` is the address of the caller Compliance contract.
+    *  `_newExchangeID` is the ONCHAINID address of the exchange to add.
+    */
+    event ExchangeIDAdded(address indexed compliance, address _newExchangeID);
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:101]:
```suggestion
    event ExchangeIDAdded(address indexed _newExchangeID);
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+    /**
+     *  this event is emitted whenever an ONCHAINID is untagged as belonging to an exchange.
+     *  the event is emitted by 'removeExchangeID'.
+     *  `compliance` is the address of the caller Compliance contract.
+     *  `_exchangeID` is the ONCHAINID being untagged as an exchange ID.
+     */
+    event ExchangeIDRemoved(address indexed compliance, address _exchangeID);
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:109]:
```suggestion
    event ExchangeIDRemoved(address indexed _exchangeID);
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+
+    error ONCHAINIDAlreadyTaggedAsExchange(address compliance, address _exchangeID);
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:111]:
```suggestion
    error ONCHAINIDAlreadyTaggedAsExchange(address _exchangeID);
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+
+    error ONCHAINIDNotTaggedAsExchange(address compliance, address _exchangeID);
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:113]:
```suggestion
    error ONCHAINIDNotTaggedAsExchange(address _exchangeID);
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+    }
+
+    /**
+    *  @dev tags the ONCHAINID as being an exchange ID
+    *  @param _exchangeID ONCHAINID to be tagged
+    *  Function can be called only by the compliance contract
+    *  Cannot be called on an address already tagged as being an exchange
+    *  emits an `ExchangeIDAdded` event
+    */
+    function addExchangeID(address _exchangeID) external onlyComplianceCall {
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:133]:
```suggestion
    function addExchangeID(address _exchangeID) external onlyOwner {
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+        if (isExchangeID(msg.sender, _exchangeID)) {
+            revert ONCHAINIDAlreadyTaggedAsExchange(msg.sender, _exchangeID);
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:135]:
```suggestion
            revert ONCHAINIDAlreadyTaggedAsExchange(_exchangeID);
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+        }
+
+        _exchangeIDs[msg.sender][_exchangeID] = true;
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:138]:
```suggestion
        _exchangeIDs[_exchangeID] = true;
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+        emit ExchangeIDAdded(msg.sender, _exchangeID);
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:139]:
```suggestion
        emit ExchangeIDAdded(_exchangeID);
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+     */
+    function setExchangeMonthlyLimit(address _exchangeID, uint256 _newExchangeMonthlyLimit) external onlyComplianceCall {
+        _exchangeMonthlyLimit[msg.sender][_exchangeID] = _newExchangeMonthlyLimit;
+        emit ExchangeMonthlyLimitUpdated(msg.sender, _exchangeID, _newExchangeMonthlyLimit);
+    }
+
+    /**
+    *  @dev tags the ONCHAINID as being an exchange ID
+    *  @param _exchangeID ONCHAINID to be tagged
+    *  Function can be called only by the compliance contract
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:129]:
```suggestion
    *  Function can be called only by the owner of this module
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+        }
+
+        _exchangeIDs[msg.sender][_exchangeID] = true;
+        emit ExchangeIDAdded(msg.sender, _exchangeID);
+    }
+
+    /**
+    *  @dev untags the ONCHAINID as being an exchange ID
+    *  @param _exchangeID ONCHAINID to be untagged
+    *  Function can be called only by the compliance contract
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:145]:
```suggestion
    *  Function can be called only by the Owner of this module
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+    *  Cannot be called on an address not tagged as being an exchange
+    *  emits an `ExchangeIDRemoved` event
+    */
+    function removeExchangeID(address _exchangeID) external onlyComplianceCall {
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:149]:
```suggestion
    function removeExchangeID(address _exchangeID) external onlyOwner {
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+        if (!isExchangeID(msg.sender, _exchangeID)) {
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:150]:
```suggestion
        if (!isExchangeID(_exchangeID)) {
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+            revert ONCHAINIDNotTaggedAsExchange(msg.sender, _exchangeID);
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:151]:
```suggestion
            revert ONCHAINIDNotTaggedAsExchange(_exchangeID);
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+        }
+        _exchangeIDs[msg.sender][_exchangeID] = false;
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:153]:
```suggestion
        delete _exchangeIDs[_exchangeID];
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+        emit ExchangeIDRemoved(msg.sender, _exchangeID);
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:154]:
```suggestion
        emit ExchangeIDRemoved(_exchangeID);
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+        }
+
+        return true;
+    }
+
+
+    /**
+    *  @dev getter for `_exchangeIDs` variable
+    *  tells to the caller if an ONCHAINID belongs to an exchange or not
+    *  @param compliance the Compliance smart contract to be checked
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:220]:
```suggestion
    
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+    *  @param _exchangeID ONCHAINID to be checked
+    *  returns TRUE if the address corresponds to an exchange, FALSE otherwise
+    */
+    function isExchangeID(address compliance, address _exchangeID) public view returns (bool){
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:224]:
```suggestion
    function isExchangeID(address _exchangeID) public view returns (bool){
```

File: contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+        return _exchangeIDs[compliance][_exchangeID];
```
Comment (Joachim-Lebrun) [inline contracts/compliance/modular/modules/ExchangeMonthlyLimitsModule.sol:225]:
```suggestion
        return _exchangeIDs[_exchangeID];
```

**Review by Joachim-Lebrun — CHANGES_REQUESTED** [review]

**Review by Joachim-Lebrun — APPROVED** [review]

**Pull request merged**